Source/AwsCommonRuntimeKit/io/HostResolver.swift

```diff
@@ -4,98 +4,93 @@
+            withUnsafePointer(to: getHostResolutionConfig()) { hostResolutionConfigPointer in
+                if aws_host_resolver_resolve_host(rawValue,
+                        hostStr.rawValue,
+                        onHostResolved,
+                        hostResolutionConfigPointer,
+                        continuationCore.passRetained()) != AWS_OP_SUCCESS {
+                    // TODO: this is wrong. Sometimes it triggers the error callback and sometimes it doesn't.
+                    // I have a fix in progress in aws-c-io
+                    continuationCore.release()
+                    continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                }
+            }
         })
     }
 
-    private func resolve(host: String, continuation: HostResolvedContinuation) {
-        let options = ResolverOptions(resolver: self,
-                                      host: AWSString(host, allocator: allocator),
-                                      continuation: continuation)
-        let pointer: UnsafeMutableRawPointer = fromPointer(ptr: options)
+    func getHostResolutionConfig() -> aws_host_resolution_config {
+        var cHostResolutionConfig = aws_host_resolution_config()
+        cHostResolutionConfig.max_ttl = maxTTL
+        cHostResolutionConfig.impl = aws_default_dns_resolve
+        return cHostResolutionConfig
+    }
 
-        if (aws_host_resolver_resolve_host(rawValue,
-                                       options.host.rawValue,
-                                       onHostResolved, config, pointer)) != AWS_OP_SUCCESS {
-            continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
-        }
+    deinit {
+        aws_host_resolver_release(rawValue)
     }
 }
 
-private func onHostResolved(_ resolver: UnsafeMutablePointer<aws_host_resolver>!,
-                            _ hostName: UnsafePointer<aws_string>!,
+private func onHostResolved(_ resolver: UnsafeMutablePointer<aws_host_resolver>?,
+                            _ hostName: UnsafePointer<aws_string>?,
                             _ errorCode: Int32,
-                            _ hostAddresses: UnsafePointer<aws_array_list>!,
+                            _ hostAddresses: UnsafePointer<aws_array_list>?,
                             _ userData: UnsafeMutableRawPointer!) {
-    let options = userData.assumingMemoryBound(to: ResolverOptions.self)
+    let hostResolverCore = Unmanaged<ContinuationCore<[HostAddress]>>.fromOpaque(userData).takeRetainedValue()
+    if errorCode != AWS_OP_SUCCESS {
+        hostResolverCore.continuation.resume(throwing: CommonRunTimeError.crtError(CRTError(code: errorCode)))
+        return
+    }
 
-    let length = aws_array_list_length(hostAddresses)
-    var addresses: [HostAddress] = Array(repeating: HostAddress(), count: length)
+    // Success
+    let length = aws_array_list_length(hostAddresses!)
+    var addresses = [HostAddress]()
```

Comment:
This looks great as is but I just want to highlight that Swift has some nice tools to write code in a more functional programming style which has the nice effect of allowing us to define `addreses` as a constant -> `let addresses`

```swift
let addresses = (0..<length).map { index in 
    var address: UnsafeMutableRawPointer! = nil
    aws_array_list_get_at_ptr(hostAddresses!, &address, index)
    let hostAddressCType = address.bindMemory(to: aws_host_address.self, capacity: 1).pointee
    return HostAddress(hostAddress: hostAddressCType)
}
```

No need to make any changes here, just wanted to highlight :)

Reply:
Thanks, updated